mlir/lib/Catalyst/Transforms/SplitMultipleTapes.cpp

```diff
@@ -0,0 +1,346 @@
+        unsigned int curTape = 0;
+
+        func->walk([&](Operation *op) {
+            if (op == func) {
+                return; // don't visit the funcop itself
+            }
+            if (isa<catalyst::quantum::DeviceInitOp>(op)) {
+                curTape++;
+            }
+            OpsEachTape[curTape].push_back(op);
+            if ((isa<catalyst::quantum::DeviceReleaseOp>(op)) &&
+                (curTape == OpsEachTape.size() - 2)) {
+                // reached post processing
+                curTape++;
+            }
+        });
+    } // collectOperationsForEachTape()
+
+    unsigned int getResultNumber(Operation *op, Value v)
+    {
+        // Get the result index i such that v = op->getResult(i)
+        for (unsigned int i = 0; i < op->getNumResults(); i++) {
+            if (op->getResult(i) == v) {
+                return i;
+            }
+        }
+        assert(false && "Value is not a result of the operation!");
+    } // getResultNumber()
+
+    void collectNecessaryValuesFromEarlierTapes(
+        const std::vector<Operation *> &TapeOps,
+        SmallSet<std::pair<Operation *, unsigned int>, SmallSetSize>
+            &NecessaryValuesFromEarlierTapes)
+    {
+        // Go through a list of operations and collect all the necessary operand values
+        // not defined in this list itself, and add them to NecessaryValuesFromEarlierTapes
+        for (Operation *PPOp : TapeOps) {
+            PPOp->walk([&](Operation *op) {
+                for (auto operand : op->getOperands()) {
+                    if (isa<BlockArgument>(operand)) {
+                        // Arguments to the original multitape function are always available
+                        // and do not have defining ops, so don't process them
+                        continue;
+                    }
+                    Operation *OperandSource = operand.getDefiningOp();
+                    unsigned int whichResult = getResultNumber(OperandSource, operand);
+                    if (std::find(TapeOps.begin(), TapeOps.end(), OperandSource) == TapeOps.end()) {
+                        // A list operand not produced in list itself, must be from earlier
+                        // tapes/preprocessing!
+                        // Note that we should not record the same value twice in the set
+
+                        // Enforce the pair as rvalue for `contains`
+                        auto &&pair = std::make_pair(OperandSource, whichResult);
+                        if (!NecessaryValuesFromEarlierTapes.contains(pair)) {
+                            NecessaryValuesFromEarlierTapes.insert(
+                                std::make_pair(OperandSource, whichResult));
+                        }
+                    }
+                }
+            });
+        }
+    } // collectNecessaryValuesFromEarlierTapes()
+
+    void getNecessaryTapeReturns(SmallVector<Value> &RetValues,
+                                 const std::vector<Operation *> &TapeOps,
```

Comment:
It might be possible to turn TapeOps into a set as well, since mostly it is used for checking membership, except one place where its `front` was used to set an insertionpoint. 

The insertion point might be able to be set from the `set TapeOps` with some sort of `set::if_find`. Let me explore this a bit.

Reply:
Ah, I no longer think this is possible. We have 

```
        for (auto op : TapeOps) {
            op->moveBefore(&executeRegionOp.getRegion().front(), it);
        }
```

So `TapeOps`'s order is actually important and we don't want `set`'s internal ordering rule to overwrite the manual order we give it (since the operations in the tape are not necessarily, and in fact almost never, in the order of their hash values' order)

Reply:
Potentially we could investigate using `set` for the `TapeOps` but not mandatory for now